[AArch64] Add an AArch64 pass for loop idiom transformations

david-arm opened the pull request:
We have added a new pass that looks for loops such as the following:

```
  while (i != max_len)
      if (a[i] != b[i])
          break;

  ... use index i ...
```

Although similar to a memcmp, this is slightly different because instead of returning the difference between the values of the first non-matching pair of bytes, it returns the index of the first mismatch. As such, we are not able to lower this to a memcmp call.

The new pass can now spot such idioms and transform them into a specialised predicated loop that gives a significant performance improvement for AArch64. It is intended as a stop-gap solution until this can be handled by the vectoriser, which doesn't currently deal with early exits.

This specialised loop makes use of a generic intrinsic that counts the trailing zero elements in a predicate vector. This was added in https://reviews.llvm.org/D159283 and for SVE we end up with brkb & incp instructions.

Although we have added this pass only for AArch64, it was written in a generic way so that in theory it could be used by other targets. Currently the pass requires scalable vector support and needs to know the minimum page size for the target, however it's possible to make it work for fixed-width vectors too. Also, the llvm.experimental.cttz.elts intrinsic used by the pass has generic lowering, but can be made efficient for targets with instructions similar to SVE's brkb, cntp and incp.

Original version of patch was posted on Phabricator:

 https://reviews.llvm.org/D158291

Patch co-authored by Kerry McLaughlin (@kmclaughlin-arm) and David Sherwood (@david-arm)

See the original discussion on Discourse:
https://discourse.llvm.org/t/aarch64-target-specific-loop-idiom-recognition/72383

david-arm committed:
[AArch64] Add an AArch64 pass for loop idiom transformations

We have added a new pass that looks for loops such as the following:

  while (i != max_len)
      if (a[i] != b[i])
          break;

  ... use index i ...

Although similar to a memcmp, this is slightly different because
instead of returning the difference between the values of the first
non-matching pair of bytes, it returns the index of the first mismatch.
As such, we are not able to lower this to a memcmp call.

The new pass can now spot such idioms and transform them into
a specialised predicated loop that gives a significant performance
improvement for AArch64. It is intended as a stop-gap solution until
this can be handled by the vectoriser, which doesn't currently deal
with early exits.

This specialised loop makes use of a generic intrinsic that counts
the trailing zero elements in a predicate vector. This was added in
https://reviews.llvm.org/D159283 and for SVE we end up with brkb &
incp instructions.

Although we have added this pass only for AArch64, it was written in
a generic way so that in theory it could be used by other targets.
Currently the pass requires scalable vector support and needs to know
the minimum page size for the target, however it's possible to make
it work for fixed-width vectors too. Also, the llvm.experimental.cttz.elts
intrinsic used by the pass has generic lowering, but can be made
efficient for targets with instructions similar to SVE's
brkb, cntp and incp.

Original version of patch was posted on Phabricator:

 https://reviews.llvm.org/D158291

Patch co-authored by Kerry McLaughlin (@kmclaughlin-arm) and David Sherwood (@david-arm)

See the original discussion on Discourse:
https://discourse.llvm.org/t/aarch64-target-specific-loop-idiom-recognition/72383

sjoerdmeijer commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,726 @@
+
+  using namespace PatternMatch;
+
+  // The incoming value to the PHI node from the loop should be an add of 1.
```
Can we use one of the Loop APIs for this? For example:

```
PHINode * getCanonicalInductionVariable () const
Check to see if the loop has a canonical induction variable: an integer recurrence that starts at 0 and increments by one each time through the loop.
```

david-arm replied:
Sadly not! That's because our induction variable does not start at 0.

sjoerdmeijer commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,726 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
Some high level descriptions what patterns are recognised would be nice.

sjoerdmeijer commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,726 @@
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aarch64-lit"
```
Bikeshedding: `lit` as name could perhaps be confusing. Something more verbose but descriptive with "loopidiom" in it would probably have my preference.

david-arm replied:
Sorry, I just realised I have lost this change somehow. I'll fix it.

sjoerdmeijer commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,726 @@
+  //   br i1 %cmp.not, label %while.end, label %while.body
+  //
+  auto CondBBInsts = LoopBlocks[0]->instructionsWithoutDebug();
+  if (std::distance(CondBBInsts.begin(), CondBBInsts.end()) > 4)
```
Counting the number of instructions might be a bit too "specific". Do need to inspect the supported instructions instead?

antoniofrighetto replied:
Is there a Godbolt link for the C leading to these instructions (I couldn't reproduce the same sequence for `compare_bytes_simple`)? I know this is to discriminate specific loop structures so chances for pattern matching later to succeed are higher, but this looks very specific to me too, and may be subject to quick changes in the future. Would dropping this lead to bottlenecks?

david-arm replied:
Sure. https://godbolt.org/z/e5MnnGT84

As you can see, the first block corresponds to this - `while (++start != end)`, so the block is likely to be tiny. It's possible the compiler could increment `start` before the loop, in which case it will look more like `while (start++ != end)`. We've seen the former idiom in xz and 7zip. In either case the first block will have 4 or less instructions. Even if we permit more instructions we still have to match the right pattern - it would likely fail anyway because we have more than the expected uses of a value. We also couldn't, for example, permit random stores in the loop to account for the extra instructions.

I'm a bit reluctant to expand the algorithm to cover hypothetical cases that we can't prove exist. I agree it's a good idea to make it robust based on real-world examples. Do you have any examples of other specific variations of this idiom that you are worried about?

david-arm replied:
It's worth noting that the LoopIdiomRecognition pass also uses std::distance as a way to filter out loops.

For what it's worth, I do have downstream code that increases the number of idioms we recognise. This led to me finding a number of bugs, so the exercise was very useful! However, I'd prefer to land this patch first and build upon this incrementally. I can create a patch later on that adds these extra idioms once I've done sufficient testing. I think the idioms in this patch have the biggest impact in terms of performance.

The extra idioms involve: 1) supporting 64-bit induction variable types, 2) permitting add increments without nuw and nsw flags, 3) supporting loops that increment the counter after loading the values (as opposed to currently incrementing before loading).

antoniofrighetto replied:
Sorry for replying only now, thanks for the explanation. If I got this correctly, the snippet you provided is not meant to be transformed as the indexes of the two GEPs do not match, thus we would be bailing out. I'm confused though, why the following snippet shouldn't be supported (as per comment [here](https://github.com/llvm/llvm-project/blob/5f09a37fc4d808310b8f0ae23f10f953d05c8749/llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp#L359-L376))? Doesn't it match the specifics of the pass (the index is simply incremented before loading the values), or am I missing something?
```c
int test(unsigned char *p1, unsigned char *p2, unsigned start, unsigned end) {
  while (++start != end) {
    if (p1[start] != p2[start])
      break;
  }
  return start;
}
```
Also, if we were to support 1), 2), 3) we would still not be catching samples like https://godbolt.org/z/vPGqde4zj, due to the `and` on the induction variable (not sure why it has been converted to an i64 here, but I guess it's an orthogonal problem – provided we want to catch this).

david-arm replied:
This logic is designed to prevent recognising loops such as `@compare_bytes_simple2`, i.e.

```
while.cond:
...
  br i1 %cmp.not, label %while.end, label %while.body

while.body:
...
  br i1 %cmp.not2, label %while.cond, label %while.end

while.end:
...
  %final_ptr = phi ptr [ %c, %while.body ], [ %d, %while.cond ]
```

where the found and end blocks are the same, and in that block there is a PHI node that has a unique incoming value for each block in the loop. However, each incoming value is not defined in the loop itself. Currently we don't handle this case in `transformByteCompare` because it would require extra work in the `byte.compare` block to select between these two incoming values. Rather than let `transformByteCompare` generate incorrect code and lead to bugs, I've explicitly disabled these cases for now. I haven't encountered any examples of such loops in the benchmarks I have looked at, so these are not critical to us. I only discovered this accidentally while writing more test cases.

sjoerdmeijer commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,726 @@
+  // At this point we know two things must be true:
+  //  1. Start <= End
+  //  2. ExtMaxLen <= 4096 due to the page checks.
+  // Therefore, we know that we can use a 64-bit induction variable that
```
I am missing some higher level descriptions of the transformation that we are exactly trying to achieve, and under which conditions this is a valid thing to do. For example, I don't understand yet the purpose of the page size check.

david-arm replied:
I've added comments at the top of the file, as well as detailed comments further up explaining how the page size is used to generate runtime memory checks. These ensure we don't fault midway through a vector load due to reading ahead.

xgupta commented:
Sorry, I was supposed to push to my repo, but changes got into this pull request, I will revert it immediately.

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,726 @@
+
```
This extra blank line can be removed.

david-arm committed:
Address review comments

* Added option to override the target's minimum page size.
* Added more comments/documentation explaining what the pass
is doing.
* Rewrote the code for detecting the indices.

david-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.h:
```diff
@@ -419,6 +419,8 @@ class AArch64TTIImpl : public BasicTTIImplBase<AArch64TTIImpl> {
     return BaseT::getStoreMinimumVF(VF, ScalarMemTy, ScalarValTy);
   }
+
+  std::optional<unsigned> getMinPageSize() const { return 4096; }
```
Good suggestion! I added this in TargetTransformInfo.cpp in a similar way to the CacheLineSize option.

david-arm committed:
Merge branch 'llvm:main' into aarch64_idiom

david-arm committed:
Fix some issues with the loop idiom recognition and loop generation

As an experiment I tried adding support for more loop idioms in.
As a result of greater test coverage I also discovered issues with
how we create the loops, which led to me fixing these bugs:

* The loops were not in the expected LCSSA format, which I have
now fixed. I've added an option for testing that allows us to
verify the loops after we create them.
* The scalar fallback loop could potentially read one byte past the
end of the array. Curiously it would always have given a correct
result, but had the potential to seg fault.
* I also added code to reject an unsupported case that would have
led to incorrect code generation. I have added a test for this.

I also amended some comments and cleaned up some unused variables.

I haven't added support for more idioms in this patch yet, since
I would like to do that in a follow-on patch after I've had chance
to do more testing. It would be good to build on this incrementally.

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,839 @@
+//
+// TODO List:
+//
+// * When optimizing for code size we may want to avoid some transformations.
```
Can we add a check for this now? It looks like it will be quite a bit bigger?

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,839 @@
+  Value *RhsPageCmp = Builder.CreateICmpNE(RhsStartPage, RhsEndPage);
+
+  Value *CombinedPageCmp = Builder.CreateOr(LhsPageCmp, RhsPageCmp);
+  BranchInst *CombinedPageCmpCmpBr = BranchInst::Create(
```
Did you consider this setting the maximum number of iteration for the new loop, as opposed to just skipping it? Or does that lead to worse performance?

david-arm replied:
So a while ago I did try writing a simple SVE ACLE implementation that processed the data in chunks, constantly walking up to the nearest page boundary in the SVE loop and then restarting again. The performance wasn't any better (presumably due to the extra calculations of the end index) and I figured it would add even more complexity to expandFindMismatch, so I stuck with this approach for now. But I'm happy to be proved wrong and this is certainly something we could revisit.

david-arm committed:
Address recent review comments

* Bail out of transformation if optimising for size
* Simple code tidy-ups
* Add support for incrementing the index with wrap flags

david-arm commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,839 @@
+            return false;
+        }
+
+  // Don't replace the loop if the add has a wrap flag.
```
We were originally being conservative, because we were worried that we'd have to rewrite the algorithm based on the presence of these flags. However, I now understand these flags are more like hints for optimisations that can choose to ignore them. I've propagated these flags into the scalar version of the loop.

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,816 @@
+  PHINode *SVEIndexPhi = Builder.CreatePHI(I64Type, 2, "mismatch_sve_index");
+  SVEIndexPhi->addIncoming(ExtStart, SVELoopPreheaderBlock);
+  Type *SVELoadType = ScalableVectorType::get(Builder.getInt8Ty(), 16);
+  Value *GepOffset = SVEIndexPhi;
```
This variable isn't needed?

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,816 @@
+  DTU.applyUpdates({{DominatorTree::Insert, SVELoopMismatchBlock, EndBlock}});
+
+  // Generate code for scalar loop.
+  Builder.SetInsertPoint(LoopPreHeaderBlock);
```
Why is the scalar loop recreated as opposed to reused? Is it the same but rotated?

david-arm replied:
Yeah, it's the same but we've already incremented the value by 1 before entering the loop. I guess that's what you mean by rotated? You're right in theory that we could have reused the existing loop, but given we wanted to increment the index before the vector loop I thought it was good to be consistent with the scalar loop as well.

davemgreen commented on llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:
```diff
@@ -0,0 +1,816 @@
+
+  // If no mismatch was found, we can jump to the end block. Create a
+  // new basic block for the compare instruction.
+  auto *CmpBB = BasicBlock::Create(Preheader->getContext(), "byte.compare",
```
Is CmpBB needed, or is this really part of the end of PreHeader, after the split?

david-arm replied:
Given the way that expandFindMismatch funnels all paths from both the scalar and vector loops into a single mismatch.end block, we then need a final byte.compare block to check whether the result is the end or not. That decides which successor block to jump to - FoundBB or EndBB. I admit for the case when FoundBB == EndBB we don't need to do this, but I think it's simpler just to be consistent with the FoundBB != EndBB case.

davemgreen replied:
I think my point was that (old) preheader branches unconditionally to CmpBB, and CmpBB branches to EndBB/FoundBB. The preheader could be branch to EndBB/FoundBB. The condition for the preheader is `Builder.CreateCondBr(Builder.getTrue(), CmpBB, Header);` though, so it might make sense to keep it as is and let simplify-cfg clean up the branches.

davemgreen reviewed (commented):
Thanks. I think it is worth trying to get this in. I already see it triggering in a number of places, it might be worth working on making it a little more generic in followup patches if we can, but there is already quite a bit going on.

david-arm committed:
Address more review comments

* Do more work to ensure DominatorTree is up to date.
* Add more checks to ensure only one use of the index PHI -
transformByteCompare doesn't currently handle this case as it
would require subtracting one from the result. It's not a
typical use case for loops in this form.
* General clean-up.
* Add the hasOptSize check back in as it was somehow lost in
an intermediate commit.

davemgreen reviewed (approved):
Thanks for the updates. From what I can tell this LGTM, but it will need a rebase.

You might want to commit it with the option disabled, and then flip the switch in a followup to avoid the commit-revert cycles in case there are any issues.

david-arm committed:
Merge branch 'llvm:main' into aarch64_idiom

david-arm committed:
Rebase and disable by default

* Fixed an issue after rebase with change in
registerPassBuilderCallbacks definition.
* Disabled the new pass by default - we can enable this in
a later patch.

dyung commented:
Hi @david-arm, the test you added, byte-compare-index.ll seems to be failing on several bots, can you take a look?

For example on https://lab.llvm.org/buildbot/#/builders/139/builds/56740:
```
******************** TEST 'LLVM :: Transforms/LoopIdiom/AArch64/byte-compare-index.ll' FAILED ********************
Exit Code: 2
Command Output (stderr):
--
RUN: at line 2: /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/opt -aarch64-lit -disable-aarch64-lit-all=false -aarch64-lit-verify -verify-dom-info -mtriple aarch64-unknown-linux-gnu -mattr=+sve -S < /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/llvm/test/Transforms/LoopIdiom/AArch64/byte-compare-index.ll | /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/FileCheck /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/llvm/test/Transforms/LoopIdiom/AArch64/byte-compare-index.ll
+ /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/FileCheck /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/llvm/test/Transforms/LoopIdiom/AArch64/byte-compare-index.ll
+ /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/opt -aarch64-lit -disable-aarch64-lit-all=false -aarch64-lit-verify -verify-dom-info -mtriple aarch64-unknown-linux-gnu -mattr=+sve -S
opt: Unknown command line argument '-aarch64-lit'.  Try: '/home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/opt --help'
opt: Did you mean '--march'?
opt: Unknown command line argument '-disable-aarch64-lit-all=false'.  Try: '/home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/opt --help'
opt: Did you mean '--disable-sched-stalls=false'?
opt: Unknown command line argument '-aarch64-lit-verify'.  Try: '/home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/opt --help'
opt: Did you mean '--tail-dup-verify'?
FileCheck error: '<stdin>' is empty.
FileCheck command line:  /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/build/bin/FileCheck /home/buildbot/buildbot-root/llvm-clang-x86_64-sie-ubuntu-fast/llvm-project/llvm/test/Transforms/LoopIdiom/AArch64/byte-compare-index.ll
--
```

david-arm commented:
Hi @dyung, sorry about this! It passed for me locally. It sounds like it needs a REQUIRED aarch64-target somewhere then.

I'll try to fix it asap.

david-arm commented:
@dyung - fix pending here https://github.com/llvm/llvm-project/pull/77467

david-arm commented:
@dyung - fix pending here https://github.com/llvm/llvm-project/pull/77467

shaojingzhi commented:
Hi! I wonder that have you conducted any tests to determine the potential performance increase of this pass in the SPEC2017 557xz benchmark? I attempted to apply it to the xz benchmark, but only one copy(--copies=1) demonstrated a significant increase(about 3%), but there was no increase when I set --copies=128 or higher. Do you have any suggestions or test results that you could share?

david-arm commented:
> Hi! I wonder that have you conducted any tests to determine the potential performance increase of this pass in the SPEC2017 557xz benchmark? I attempted to apply it to the xz benchmark, but only one copy(--copies=1) demonstrated a significant increase(about 3%), but there was no increase when I set --copies=128 or higher. Do you have any suggestions or test results that you could share?

The most significant gains with xz have already been achieved when https://github.com/llvm/llvm-project/pull/77480 and https://github.com/llvm/llvm-project/pull/77480 landed, which improved performance by 6-7% for neoverse-v1. This PR is a NFC refactoring patch so it won't improve performance further. My follow-on patch (not yet posted) will trigger more cases in xz, but I don't expect any substantial performance gains for xz. The main goal of extending this pass further is to improve code coverage and testing, and hopefully there will be other applications besides xz that will benefit too.
